onnxruntime/contrib_ops/cpu/activations.h

```diff
@@ -73,5 +73,68 @@ class Gelu : public OpKernel {
   }
 };
 
+template <typename T>
+class FastGelu : public OpKernel {
+ public:
+  FastGelu(const OpKernelInfo& info) : OpKernel(info) {}
+
+  Status Compute(OpKernelContext* context) const override {
+    const auto* X = context->Input<Tensor>(0);
+    Tensor* Y = context->Output(0, X->Shape());
+    concurrency::ThreadPool* tp = context->GetOperatorThreadPool();
+    if (nullptr != tp) {
+      const T* input = X->template Data<T>();
+      T* output = Y->template MutableData<T>();
+      int task_count = tp->NumThreads() + 1;
+      int64_t elem_count = X->Shape().Size();
+      const auto coefficient = kAlpha * kGamma;
+      if (elem_count > task_count) {
+        tp->ParallelFor(task_count, [ input,
+                                      output,
```

Comment:
Would be better if the number of threads used was relative to the amount of work to do. e.g. if there are 8 threads and 8 elements do you really want to process each individual element on a different thread or is there a minimum number of elements per thread that would make more sense?

Reply:
Are you suggesting creating different number of threads according to the work load? Creating thread could be expensive and thus having 8 elements on 8 threads maybe less performance than having one thread running 8 elements. But I think the threads here are pre-created and parked in a thread pool, so we just go and grab it. In this case, I think it's ok to have each thread processing individual elements.

Reply:
Yes. Whilst the threads are already created, there is still overhead from using them (data has to move to different cores, waiting on the threads to all finish as they will start/end at different times) so you are most likely slowing down things by using all threads when there aren't many elements to process. Would be good to at least come up with a rough metric of min elements per thread and use that when deciding how many threads to use.

---
In reply to: [400608625](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r400608625) [](ancestors = 400608625)

Reply:
Mark the comment as resolved. I think in the newer version of FastGelu, Tianlei selected N = 4096 as the number of elements per thread based on perf test. 

---
In reply to: [404408889](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r404408889) [](ancestors = 404408889,400608625)